### PR TITLE
DM-4591: GWT conversion: System notifications

### PR DESCRIPTION
**Pull request by loitly**
ticket: https://jira.lsstcorp.org/browse/DM-4591

to test:
rebuild firefly.  restart tomcat.
http://localhost:8080/firefly/
mkdir -p /hydra/alerts/
echo 'some message you want users to see' > /hydra/alerts/notice.txt
within 10 seconds, you should see the notice appears on the drop down menu.
remove the file, and it'll go away.

### CODE REVIEW

File: src/firefly/java/edu/caltech/ipac/firefly/server/AlertsMonitor.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+ * Date: 6/16/16
+ *
+ * @author loi
+ * @version $Id: $
+ */
+public class AlertsMonitor {
+    private static final Logger.LoggerImpl LOG = Logger.getLogger();
+    private static final String ALERTS_DIR = AppProperties.getProperty("alerts.dir", "/hydra/alerts/");
+    private static final File alertDir = new File(ALERTS_DIR);
+    private static Alert alerts = new Alert(null, "", true);;
```
Comment (tgoldina) [inline src/firefly/java/edu/caltech/ipac/firefly/server/AlertsMonitor.java]: extra semicolon

File: src/firefly/java/edu/caltech/ipac/firefly/server/AlertsMonitor.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
+package edu.caltech.ipac.firefly.server;
+
+import com.google.gwt.media.dom.client.TimeRanges;
```
Comment (tgoldina) [inline src/firefly/java/edu/caltech/ipac/firefly/server/AlertsMonitor.java]: looks like a wrong import on server side, a leftover?

File: src/firefly/java/edu/caltech/ipac/firefly/server/AlertsMonitor.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    private static Timer timer;
+
+    /**
+     * check for alerts
+     * @param forceSend  if true, send alerts regardless
+     */
+    public static void checkAlerts(boolean forceSend) {
+        long lastMod = alerts == null ? 0 : alerts.getLastModDate();
+        if (alertDir.lastModified() > lastMod) {
+            File[] files = alertDir.listFiles(file -> file.isFile() && !file.isHidden());
```
Comment (tgoldina) [inline src/firefly/java/edu/caltech/ipac/firefly/server/AlertsMonitor.java]: Nice use of lambda expression in java! We should all start using them more!

File: src/firefly/java/edu/caltech/ipac/firefly/server/AlertsMonitor.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        timer = new Timer("AlertsMonitor", true);
+        timer.scheduleAtFixedRate(new TimerTask() {
+            public void run() {
+                checkAlerts(false);
+            }
+        }, 0, 1000 * 10);  // check for update every 10 seconds
+    }
+
+    private static void sendAlerts(Alert alerts) {
+        FluxAction action = new FluxAction("app_data.setAlerts");       // AppDataCntlr.SET_ALERTS
```
Comment (tgoldina) [inline src/firefly/java/edu/caltech/ipac/firefly/server/AlertsMonitor.java]: Good to know we can fire flux action from server side. Somehow it missed it before.

File: src/firefly/java/edu/caltech/ipac/firefly/server/RequestOwner.java [hunk 5]
```diff
@@ -64,15 +64,17 @@ public RequestAgent getRequestAgent() {
...
-        host = requestAgent.getHeader("host");
-        referrer = requestAgent.getHeader("Referer");
-
-        String sei = requestAgent.getCookie("seinfo");
-        if (!StringUtils.isEmpty(sei)) {
-            String [] parts =  sei.split("_");
-            if (parts.length > 1) {
-                eventConnID = parts[0];
-                eventChannel = parts[1];
+        if (requestAgent != null) {
```
Comment (tgoldina) [inline src/firefly/java/edu/caltech/ipac/firefly/server/RequestOwner.java]: just curious: when the requestAgent would be null?

File: src/firefly/js/ui/DropDownContainer.jsx [hunk 6]
```diff
@@ -84,20 +86,25 @@ export class DropDownContainer extends Component {
...
-                <div className='DD-ToolBar__content'>
-                    {view}
-                </div>
-                <div id='footer' className='DD-ToolBar__footer'>
-                    {footer}
-                    <div className='DD-ToolBar__version'>
-                        {getVersion()}
+            <div>
+                <div className='DD-ToolBar'>
+                    {alerts || <Alerts />}
```
Comment (tgoldina) [inline src/firefly/js/ui/DropDownContainer.jsx]: Are the alerts sent via properties will be user alerts? Is it right to display one or the other?

**Comment by tgoldina** [comment]
Looks good, though modifying the existing notice does not work.

Test sequence: 
echo 'something' > /hydra/alerts/notice.txt
   wait for message to show up
echo 'something else' > /hydra/alerts/notice.txt

The updated message never shows up.

File: src/firefly/java/edu/caltech/ipac/firefly/server/RequestOwner.java (hunk 5 above)
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/firefly/server/RequestOwner.java]: When RequestOwner is referenced outside of a http request thread.  Somewhere in the our cache code, it's doing it.

File: src/firefly/js/ui/DropDownContainer.jsx (hunk 6 above)
Comment (loitly) [inline src/firefly/js/ui/DropDownContainer.jsx]: alerts from property is meant to replace the default <Alerts> panel.  that way, one can change its look and feel.

**Pull request merged**